### PR TITLE
Improve certificate authentication

### PR DESCRIPTION
**Pull request by Firgeis**
This PR improves client certificate handling in piaf with the following:

- Allows to check server certificate validity (either from a file or from a certificate string)
- Allows to present a client certificate for authentication (from a certificate string and private key string)

Outstanding issues:

- Requires a modifiied ocaml-ssl library: https://github.com/savonet/ocaml-ssl/pull/71
- Testing for a non existing client certificate causes the test server to crash

### CODE REVIEW

**Comment by anmonteiro** [comment]
Whoa! I really like this. Thanks for putting in the work.

Give me some time to review it. 

What are some blockers for you right now?

**Comment by Firgeis** [comment]
Well mostly getting the changes to ocaml-ssl in and updating the dependency

**Comment by Firgeis** [comment]
These fixes the outstanding test issues

**Comment by anmonteiro** [comment]
Thanks, I'm gonna take a look at this soon. I pushed a commit bumping the Nix package sources to a commit that includes your openssl PR.

Let's see what CI says.

File: bin/carl.ml [hunk 1]
```diff
@@ -544,10 +544,23 @@ module CLI = struct
...
+
+  let cacertpem = 
+    let cert_conv =
+      let parse s = Ok (Cert.Certpem s) in
+      Arg.conv ~docv:"method" (parse, Cert.pp_print_cert)
+    in
     let doc = "CA certificate to verify peer against" in
     let docv = "file" in
-    Arg.(value & opt (some string) None & info [ "cacert" ] ~doc ~docv)
+    Arg.(value & opt (some cert_conv) None & info [ "cacert" ] ~doc ~docv)
```
Comment (anmonteiro) [inline bin/carl.ml:563]: are you sure these two options behave correctly by themselves, and differently than one another?

My mental model of cmdliner says only one of them will be parsed, as they're both `--cacert`

File: lib/cert.ml [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+ *---------------------------------------------------------------------------*)
+
+ type t = 
+  | Empty
+  | Filepath of string
+  | Certpem of string
+  
+ let pp_print_cert ppf = function
+  | Empty -> Format.pp_print_string ppf "Empty" 
+  | Filepath s ->  Format.pp_print_string ppf s
```
Comment (anmonteiro) [inline lib/cert.ml:39]: if this is for human debugging, can we:

1. name it `pp` or `pp_hum`
2. Add the tag information to both `Filepath` and `Certpem`, please?

File: lib/cert.ml [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *---------------------------------------------------------------------------*)
+
+ type t = 
+  | Empty
```
Comment (anmonteiro) [inline lib/cert.ml:33]: can we avoid this `Empty` variant here? IMO `Cert.t` shouldn't care about presence. I believe that's a job for the `option` type.

Unless there's a specific meaning to `Empty`, or otherwise `Some Empty` and `None` mean different things?

File: lib/openssl.ml [hunk 4]
```diff
@@ -102,6 +102,12 @@ let log_cert_info ~allow_insecure ssl_sock =
         (pp_cert_verify_result ~allow_insecure)
         verify_result)
 
+let load_client_cert cert privkey ctx =
```
Comment (anmonteiro) [inline lib/openssl.ml:105]: nit: can `cert` and `privkey` become labeled arguments here?

File: lib/openssl.ml [hunk 5]
```diff
@@ -102,6 +102,12 @@ let log_cert_info ~allow_insecure ssl_sock =
...
+  Ssl.use_certificate_from_string ctx cert privkey
+
+let load_peer_ca_cert cert ctx =
```
Comment (anmonteiro) [inline lib/openssl.ml:108]: same as the above for `cert`.

File: lib/openssl.ml [hunk 6]
```diff
@@ -293,10 +300,30 @@ let connect ~hostname ~config ~alpn_protocols fd =
...
+          
+          (* Send client cert if present *)
+          let clientcertio = match clientcert with
+            | Some certwithkey -> 
+              let cert, privkey = certwithkey in
+              Lwt_result.return (load_client_cert cert privkey ctx)
+            | None -> Lwt_result.return ()
+          in
+          
+          Lwt.bind certio (fun _ -> clientcertio)
```
Comment (anmonteiro) [inline lib/openssl.ml:322]:
```suggestion
          (* Server certificate verification *)
          let* certio = 
            match cacert with 
            | Empty -> configure_verify_locations ctx None capath
            | Filepath path -> 
              let somepath = Some(path) in
              configure_verify_locations ctx somepath capath
            | Certpem cert -> Lwt_result.return (load_peer_ca_cert cert ctx)
          in
          
          (* Send client cert if present *)
          match clientcert with
           | Some certwithkey -> 
              let cert, privkey = certwithkey in
              Lwt_result.return (load_client_cert cert privkey ctx)
            | None -> Lwt_result.return ()
```

Comment (anmonteiro) [reply lib/openssl.ml:322]: I believe the suggestion above is equivalent to your current code.

File: lib_test/helper_server.ml [hunk 7]
```diff
@@ -136,9 +139,14 @@ end
...
-let listen ?(http_port = 8080) ?(https_port = 9443) () =
+let listen ?(http_port = 8080) ?(https_port = 9443) ?(check_client_cert=false) () =
   let http_server = HTTP.listen http_port in
-  let https_server = ALPN.https_server https_port in
+  let https_server =
+    if check_client_cert then
+      ALPN.https_server https_port ~check_client_cert:true
+    else
+      ALPN.https_server https_port
+  in
```
Comment (anmonteiro) [inline lib_test/helper_server.ml:149]:
```suggestion
  let https_server = ALPN.https_server https_port ~check_client_cert in
```

File: lib_test/test_client.ml [hunk 8]
```diff
@@ -168,6 +168,111 @@ let test_https _ () =
...
+        }
+      (Uri.of_string "https://localhost:9443")
+    in
+  let response = Result.get_ok response in
+  Alcotest.check
+    response_testable
+    "expected response"
+    (Response.create 
+      ~version:Versions.HTTP.v1_1 
+      ~headers:Headers.(of_list [ Well_known.content_length, "1" ])
```
Comment (anmonteiro) [inline lib_test/test_client.ml:192]: is content length necessary here if you never read the response body in the test? Could it just be 0?

**Review by anmonteiro — COMMENTED** [review]
This looks great!

I left a couple comments that are mostly nits, some questions, and a couple code suggestions. Let me know what you think.

File: lib_test/test_client.ml (hunk 8 above)
Comment (Firgeis) [reply lib_test/test_client.ml:192]: Yes, response_testable checks the headers and it's returning 1 in content_length

File: bin/carl.ml [hunk 9]
```diff
@@ -544,10 +544,23 @@ module CLI = struct
...
 
-  let cacert =
+  let cacertfile =
+    let cert_conv =
+      let parse s = Ok (Cert.Filepath s) in
+      Arg.conv ~docv:"method" (parse, Cert.pp)
+    in
+    let doc = "CA certificate to verify peer against" in
+    let docv = "file" in
+    Arg.(value & opt (some cert_conv) None & info [ "cacertfile" ] ~doc ~docv)
```
Comment (anmonteiro) [inline bin/carl.ml:554]: I thought I'd written something down in a comment, but I can't find it now, so I might have deleted before submitting.

My motivation for `carl` was to follow `curl`'s command line API as close as possible. Unfortunately they only have a `--cacert` option, and these would not be in line with that API. Do you think we could revert this, and only support the certfile? I'ts a little cumbersome to write an entire PEM string in the command line, anyway.

File: lib/cert.ml [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *---------------------------------------------------------------------------*)
+
+ type t =
```
Comment (anmonteiro) [inline lib/cert.ml:32]: This looks much cleaner now!

**Review by anmonteiro — COMMENTED** [review]
Alright, 2nd round! Everything looks great, there's only the change in `bin/carl.ml` that I'd rather revert.

**Comment by Firgeis** [comment]
I have added the possibility of passing the cert/key pair as files, and added to carl an implementation resembling curl of client certificates

**Comment by anmonteiro** [comment]
Thank you!

File: lib_test/helper_server.ml [hunk 11]
```diff
@@ -98,17 +98,20 @@ module ALPN = struct
...
     let listen_address = Unix.(ADDR_INET (inet_addr_loopback, port)) in
+    let ca = "./certificates/ca.pem" in
     let cert = "./certificates/server.pem" in
     let priv_key = "./certificates/server.key" in
     Lwt_io.establish_server_with_client_socket
       listen_address
       (fun client_addr fd ->
         let server_ctx = Ssl.create_context Ssl.TLSv1_3 Ssl.Server_context in
         Ssl.disable_protocols server_ctx [ Ssl.SSLv23; Ssl.TLSv1_1 ];
+        Ssl.load_verify_locations server_ctx ca "";
```
Comment (anmonteiro) [inline lib_test/helper_server.ml:112]: @Firgeis why did you add this here? Trying to run the tests locally, it seems like openssl is not happy with it.

Comment (Firgeis) [reply lib_test/helper_server.ml:112]: That loads the ca certificate so it can verify the client cert properly. What error are you getting?

Comment (anmonteiro) [reply lib_test/helper_server.ml:112]: It was my bad. `dune runtest` works but `dune exec lib_test/test_client.exe` couldn't find the certificates. 

I'm exploring how to make both work with dune-sites

Comment (Firgeis) [reply lib_test/helper_server.ml:112]: Probably related to the relative paths

Comment (anmonteiro) [reply lib_test/helper_server.ml:112]: I fixed the paths and now I'm getting an infinite loop running the client certs test. Similar to what I see in #94 (CI just hangs)

Comment (Firgeis) [reply lib_test/helper_server.ml:112]: Retrieving the client certs here also uses relative paths
https://github.com/Firgeis/piaf/blob/fd5e07ceea0ef9d072abc1b39d39229c1300b85f/lib_test/test_client.ml#L225

Comment (anmonteiro) [reply lib_test/helper_server.ml:112]: That's not the issue. The problem is that the server never returns if it can't verify the peer. So the try / catch around the client request is useless because the server never really returns a response.

Comment (Firgeis) [reply lib_test/helper_server.ml:112]: Strange, because when I run the following test: https://github.com/anmonteiro/piaf/blob/6e1b433a09caf5a6dc0ae21c0ab2b1b0d5e1fe69/lib_test/test_client.ml#L278 (which is to cover that case when the client doesn't send the client cert and the server verifies it) is passing, at least when running dune test

Comment (Firgeis) [reply lib_test/helper_server.ml:112]: My suggestion was a try block around the reading of the certs not the client request, sorry if I was unclear
